Has one: direction option

lgebhardt opened the pull request:
Some background: In JR most has_one associations are really rails belongs_to associations. In the early JR releases this didn't matter and we choose to simplify the resources. However some recent changes have led to the code needing to know the underlying association type. The most common one to one association is really a belongs_to (where the foreign key is on the resource where the association is being declared).

This is an alternative proposal to #305, and intended to fix #300. Instead of requiring most `has_one` associations to be renamed to `belongs_to`, a `foreign_key_on` option is used to indicate that the foreign resource needs to be looked up to retrieve the identifier. In reviewing #305 with fresh eyes I have found we really should, for consistency, have separate operations and callbacks for updating the has_one vs belongs_to associations. But I don't think that level of granularity improves the code base. But without it it seems incomplete.

#305 style:
```ruby
class PostResource < JSONAPI::Resource
  attribute :title, :body
  belongs_to :author
  has_one :post_details
end
```

This PR's style

```ruby
class PostResource < JSONAPI::Resource
  attribute :title, :body
  has_one :author
  has_one :post_details, foreign_key_on: :related_resource
end
```

I'm favoring this alternative since it requires the least changes to peoples' existing code. There are few differences between a has_one and a belongs_to as far as JR is concerned. Since has_one associations are not that common it seems the foreign_key_on flag is the least intrusive way to make this work. 

The main downside is the associations will continue to not match the underlying model declarations. 

Opinions are welcome.

barelyknown commented:
I think that it's smart that you're thinking about minimizing the amount of change required for users, but I'm not sure that it's worth it in this case. New Rails programmers often have trouble understanding the difference between `has_one` and `belongs_to` relationships, and I'm afraid that if we don't match Rails convention here it'll confuse users more than the avoiding the pain of migrating to the new syntax is worth.

barelyknown commented:
![975-x-696](https://cloud.githubusercontent.com/assets/693652/8674069/66721e4e-2a09-11e5-9806-6328eba65172.jpg)

lgebhardt commented:
@barelyknown Thanks for your opinion (and for pointing out my poorly named branch - what was I thinking?). 

If we go to separate `belongs_to` and `has_one` associations do you see us also needing separate operations and callbacks? I'm tempted to keep those lumped together under has_one, but maybe that's shortsighted.

barelyknown commented:
My hunch is that we should have separate operations and callbacks, but I haven't come across that specific need.

lgebhardt commented:
After chatting with @dgeb I've changed the option name.

dgeb commented:
In the pre-1.0 days, I don't mind introducing breaking changes if it improves the API. In this case, I actually prefer using `has_one` to cover both scenarios, since the location of the foreign key is an implementation detail that's irrelevant to how a relationship is displayed at the API layer. Consider that resources don't have to be backed by ActiveRecord or even a relational DB.

I'm ok with the `foreign_key_on` option, but open to suggestions (other than `has_one: direction` :stuck_out_tongue_winking_eye:)

barelyknown commented:
`foreign_key_on: :related_resource` seems good to me.

ggordon commented:
What do you think about using a single method like `association` instead of  the `belongs_to`, `has_one` or `has_many` methods?

```
belongs_to -> association :author, multiplicity: :one
has_one    -> association :priority, multiplicity: :one, foreign_key_on: :priority
has_many   -> association :comments, multiplicity: :many
```
This could be further simplified if multiplicity defaulted to :one. Not sure if `association` or `multiplicity` are the best names, but the idea is to just have one method instead of two/three.

lgebhardt commented:
@ggordon I think that might be a step too far. I like the syntax of has_one vs has_many, if only because I'm familiar with it. I suspect that will be case for most Rails developers. 

However if we were to rename the methods we would probably want to adopt the naming as it's used in the JSON API specification. That would be "relationship" instead of "association".

I'm not sure I like this, but how about the following?

```
belongs_to -> relationship :author, to: :one
has_one    -> relationship :priority, to: :one, foreign_key_on: :priority
has_many   -> relationship :comments, to: :many
```

dgeb commented:
I don't like the verbosity of `:related_resource`, and I don't like having to repeat the relationship name.

My preference is to go with `foreign_key_on: :related` and `foreign_key_on: :self`

dgeb commented:
I could see supporting both `has_one` and `has_many` as aliases to `relationship`, which would be the most general form.

lgebhardt commented:
Ok, I'll make those changes.

dgeb commented:
Thanks @lgebhardt!

lgebhardt commented:
I added the `relationship` method and renamed the `foreign_key_on` option to `related`.

The last commit is a large set of changes. I used this as an opportunity to refactor "associations" into "relationships". I've been wanting to do this since JSON API started using the relationships top level key. I expect a bit of pain for people with overrides of some resource methods and callbacks. I think it improves the understandability of the project, and in the long run it should pay off. I hope you all agree.

hinshun commented on test/unit/serializer/link_builder_test.rb:
```diff
@@ -98,11 +98,11 @@ def test_relationships_self_link_for_regular_app
     builder       = JSONAPI::LinkBuilder.new(config)
     source        = Api::V1::PersonResource.new(@steve)
-    association   = JSONAPI::Association::HasMany.new("posts", {})
+    relationship   = JSONAPI::Relationship::ToMany.new("posts", {})
```
the '='s across this file are misaligned now, a small style issue but figure i'll just point it out in case you weren't aware.

lgebhardt replied:
Thank you. I hadn't noticed. I'll get it cleaned up.

dgeb commented:
Fantastic job @lgebhardt :shipit:
